cmd/proxy-server/main.go

```diff
@@ -540,3 +563,152 @@ func rolesHandler(log *logrus.Entry, opaHost string) http.Handler {
+				}
+				return
+			}
+			// Check if the tenant is being denied.
+			ok, err := rdb.SIsMember(keyTenantRevoked, claims.Group).Result()
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				if err := web.JSONErrorResponse(w, err); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+			if ok {
+				w.WriteHeader(http.StatusUnauthorized)
+				if err := web.JSONErrorResponse(w, fmt.Errorf("tenant is revoked")); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+			log.Info("Calling into roles!")
+
+			// Gather storage pool, systemtype and systemID
+
+			if roleServ.roleService == nil {
+				resp, err = roleServ.roleClient.List(r.Context(), &pb.RoleListRequest{})
+			} else {
+				resp, err = roleServ.roleService.List(r.Context(), &pb.RoleListRequest{})
+			}
+
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				if err := web.JSONErrorResponse(w, err); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+
+			roleJSON := roles.NewJSON()
+			err = roleJSON.UnmarshalJSON(resp.Roles)
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				if err := web.JSONErrorResponse(w, err); err != nil {
+					log.WithError(err).Println("sending json response")
+				}
+				return
+			}
+
+			matches := []roles.Instance{}
+			rolesSplit := strings.Split(claims.Roles, ",")
+
+			roleJSON.Select(func(rInst roles.Instance) {
+				for _, role := range rolesSplit {
+					if rInst.Name == role {
+						matches = append(matches, rInst)
+						log.Info("found role!")
+						sysID = rInst.SystemID
+						storPool = rInst.Pool
+						sysType = rInst.SystemType
+						tenant = claims.Group
+
+						dataKey := fmt.Sprintf("quota:%s:%s:%s:%s:data", sysType, sysID, storPool, tenant)
+
+						res, err := rdb.HGetAll(dataKey).Result()
+						if err != nil || len(res) == 0 {
+							w.WriteHeader(http.StatusInternalServerError)
+							if err := web.JSONErrorResponse(w, fmt.Errorf("no volumes found for tenant %s", tenant)); err != nil {
+								log.WithError(err).Println("sending json response")
+							}
+							return
+						}
+
+						for volKey := range res {
+							if strings.Contains(volKey, "capacity") {
+								splitStr := strings.Split(volKey, ":")
+								//example : vol:k8s-cb89d36285:capacity
+								if len(splitStr) == 3 {
+									volumeMap[splitStr[1]] = splitStr[1]
+								}
+							}
+							if strings.Contains(volKey, "deleted") {
+								splitStr := strings.Split(volKey, ":")
+								//example : vol:k8s-cb89d36285:deleted
+								if len(splitStr) == 3 {
+									delete(volumeMap, splitStr[1])
+								}
+							}
+						}
+					}
+				}
+			})
+
+			if len(matches) == 0 {
+				w.WriteHeader(http.StatusInternalServerError)
+				if err := web.JSONErrorResponse(w, fmt.Errorf("this role %s doesn't exist", claims.Roles)); err != nil {
```

Comment:
`claims.Roles` can have multiple roles. In that case, this message doesn't read well. Maybe something like:

`Couldn't find any configured roles matching %s`

Reply:
fixed